### PR TITLE
added tracing to matadata api

### LINKED ISSUE #4132: Add tracing for federated metadata API
### Describe the solution you'd like

Instrument tracing spans for the federated APIs. This should include the proxy (querier) side as well as the leaf store sides.

### PR DESCRIPTION
**Pull request by KrishnaSindhur**
Signed-off-by: KrishnaSindhur <krishnasindhur@gmail.com>

* [ ] I added CHANGELOG entry for this change.
* [ ] Change is not relevant to the end user.

## Changes

Add tracing for federated metadata API.

Closes #4132 

## Verification

- make build
- make test-local

### CODE REVIEW

**Comment by KrishnaSindhur** [comment]
@yeya24 can you please review this PR

File: pkg/metadata/prometheus.go [hunk 1]
```diff
@@ -31,6 +33,13 @@ func (p *Prometheus) MetricMetadata(r *metadatapb.MetricMetadataRequest, s metad
...
-		Metadata: metadatapb.FromMetadataMap(md)}})
+	tracing.DoInSpan(s.Context(), "send_metadata_response", func(_ context.Context) {
+		err = s.Send(&metadatapb.MetricMetadataResponse{Result: &metadatapb.MetricMetadataResponse_Metadata{
+			Metadata: metadatapb.FromMetadataMap(md)}})
+	})
+	if err != nil {
+		return err
+	}
+
+	return nil
```
Comment (kakkoyun) [inline pkg/metadata/prometheus.go:44]:
```suggestion
	return err
```

File: pkg/api/query/v1.go [hunk 2]
```diff
@@ -925,6 +925,15 @@ func NewMetricMetadataHandler(client metadata.UnaryClient, enablePartialResponse
 	}
 
 	return func(r *http.Request) (interface{}, []error, *api.ApiError) {
+		span, ctx := tracing.StartSpan(r.Context(), "receive_http_request")
```
Comment (kakkoyun) [inline pkg/api/query/v1.go:928]:
```suggestion
		span, ctx := tracing.StartSpan(r.Context(), "metadata_http_request")
```

File: pkg/metadata/metadata.go [hunk 3]
```diff
@@ -32,6 +33,9 @@ func NewGRPCClient(ts metadatapb.MetadataServer) *GRPCClient {
 }
 
 func (rr *GRPCClient) MetricMetadata(ctx context.Context, req *metadatapb.MetricMetadataRequest) (map[string][]metadatapb.Meta, storage.Warnings, error) {
+	span, ctx := tracing.StartSpan(ctx, "metadata_request")
```
Comment (kakkoyun) [inline pkg/metadata/metadata.go:36]:
```suggestion
	span, ctx := tracing.StartSpan(ctx, "metadata_grpc_request")
```

File: pkg/metadata/proxy.go [hunk 4]
```diff
@@ -87,7 +95,15 @@ type metricMetadataStream struct {
 }
 
 func (stream *metricMetadataStream) receive(ctx context.Context) error {
-	metadataCli, err := stream.client.MetricMetadata(ctx, stream.request)
+	var (
+		err         error
+		metadataCli metadatapb.Metadata_MetricMetadataClient
+	)
+
+	tracing.DoInSpan(ctx, "receive_stream_request", func(ctx context.Context) {
```
Comment (kakkoyun) [inline pkg/metadata/proxy.go:103]:
```suggestion
	tracing.DoInSpan(ctx, "receive_metadata_stream_request", func(ctx context.Context) {
```

**Review by kakkoyun — COMMENTED** [review]
Looking great! Thanks for the contribution.

I have added some comments inline :)

File: pkg/metadata/prometheus.go [hunk 5]
```diff
@@ -31,6 +33,13 @@ func (p *Prometheus) MetricMetadata(r *metadatapb.MetricMetadataRequest, s metad
...
 
-	return s.Send(&metadatapb.MetricMetadataResponse{Result: &metadatapb.MetricMetadataResponse_Metadata{
-		Metadata: metadatapb.FromMetadataMap(md)}})
+	tracing.DoInSpan(s.Context(), "send_metadata_response", func(_ context.Context) {
+		err = s.Send(&metadatapb.MetricMetadataResponse{Result: &metadatapb.MetricMetadataResponse_Metadata{
+			Metadata: metadatapb.FromMetadataMap(md)}})
+	})
+	if err != nil {
+		return err
+	}
```
Comment (kakkoyun) [inline pkg/metadata/prometheus.go:42]: I think you misinterpreted my previous suggestion, you should remove this block.
```suggestion
```

**Review by kakkoyun — CHANGES_REQUESTED** [review]
Nearly there. One last wrinkle to iron out.

File: pkg/metadata/prometheus.go (hunk 5 above)
Comment (kakkoyun) [reply pkg/metadata/prometheus.go:42]: @KrishnaSindhur the issue hasn't resolved so marking it unresolved. You don't need to check if err is nil.

Comment (KrishnaSindhur) [reply pkg/metadata/prometheus.go]: ya updated

**Review by kakkoyun — APPROVED** [review]

**Pull request merged**